couchdbkit/client.py

```diff
@@ -509,30 +535,28 @@ def save_doc(self, doc, encode_attachments=True, force_update=False,
             doc1['_attachments'] = resource.encode_attachments(doc['_attachments'])
 
         if '_id' in doc1:
-            docid = doc1['_id']
-            docid1 = resource.escape_docid(doc1['_id'])
+            docid = doc1['_id'].encode('utf-8')
+            couch_doc = Document(self.cloudant_database, docid)
+            couch_doc.update(doc1)
             try:
-                res = self.res.put(docid1, payload=doc1,
-                        **params).json_body
-            except ResourceConflict:
+                couch_doc.save()
```

Comment:
I just found a weird behavior change.

Prior to this change, if two processes open the same document, and concurrently one performs a save and one performs a delete, the delete operation always wins.
With this change, there exists an interleaving of operations where the save wins.  You can reproduce this by loading the same doc in two different processes, and then first delete one doc, and then second save the other doc.

The cause of this issue is in how the cloudant library implements save:
https://github.com/cloudant/python-cloudant/blob/2.7.0/src/cloudant/document.py#L181-L182

If the document is not found, we create the document.  In order to follow the previous behavior, we would need to atomically check for a deleted document and write if not deleted.

Maybe we can just call this part of Document.save to get similar behavior:
https://github.com/cloudant/python-cloudant/blob/2.7.0/src/cloudant/document.py#L184-L188

Reply:
I think I have a solution to this

Reply:
https://github.com/dimagi/couchdbkit/pull/44